### DISCUSSION TITLE
In TileMapLayer / TileSet, add support for deterministic terrain randomization

### DISCUSSION THREAD

**Discussion by c-lipka** [godotengine/godot-proposals#10948] [2D]
The TileMapLayer / TileSet terrains mechanism allows for randomization of the tiles actually picked for any given terrain, both when painting and when programmatically setting the terrain of cells via `set_cells_terrain_connect()` or `set_cells_terrain_path()`, by defining multiple tiles with the same combination of terrain and peering bits.

I want to make use of this feature from GDScript to build levels programmatically, but with the added constraint that while the levels should use some randomness in the choice of the actual tiles, this randomness should still be deterministic, i.e. whenever I re-build a level, it should look exactly the same.

I think this could be implemented by taking the random value not from a sequential PRNG (currently `Math::random()`) in `get_random_tile_from_terrains_pattern()`), but rather - either always or as an option - from a function (to be called by `set_cells_terrain_connect()` or `set_cells_terrain_path()` respectively) taking as parameters the cell coordinates and, for good measure, some (optional) user-supplied new parameter to both methods.

The function to generate the pseudorandom number could be as simple as a hash of the three parameters or a lookup into precomputed blue noise, or as sophisticated as a user-supplied functor. The latter would enable the user to have even more control over the distribution of tiles: A Perlin noise function would cause rare tiles to have a higher chance to appear in clusters, while a blue noise function would cause rare tiles to be spread more randomly; and if so desired, in this manner even geometric patterns would be possible to auto-generate.

The purpose of the third parameter, to be supplied by the user, would be to effectively choose between different distribution patterns; by default, a constant value would be used, yielding the same distribution pattern for all map layers (presuming identical probability settings) and terrains; setting the value to a hash of the terrain set and terrain ID would ensure different distribution patterns for different terrains; setting the value to the current level ID would give a different pattern for each level, and so forth.

**Comment by Calinou** [comment]
Out of curiosity, I wonder why a fixed seed isn't always being used in the first place. Exposing a way to use custom seeds or RNG functions would come later.

cc @groud

**Reply by groud** [reply]
> Out of curiosity, I wonder why a fixed seed isn't always being used in the first place. Exposing a way to use custom seeds or RNG functions would come later.

Mainly because, in the editor, if you paint an area and the pattern you got isn't great, you can just paint over it and you would have another pattern.

I think being able to provide a seed would be fine, but I don't want to complexify things a lot more, as I think a plugin/some runtime code could be used to replace the tiles however you see fits too.

**Comment by utopiamachina** [comment]
I had the same problem in our upcoming "Hunters In The Dark" game, but found a simple workaround for this. Just call `seed(...)` somewhere before calling `set_cells_terrain_connect`. 

It works and affects Math:random function that is called by `set_cells_terrain_connect`.

See Godot source code, where the cell from the terrain is chosen:
- `set_cells_terrain_connect` calls `get_random_tile_from_terrains_pattern` https://github.com/godotengine/godot/blob/65e73b3a5eec19e3269809b0f00e73a646723067/scene/2d/tile_map_layer.cpp#L3052
- `get_random_tile_from_terrains_pattern` calls `Math::random`: https://github.com/godotengine/godot/blob/65e73b3a5eec19e3269809b0f00e73a646723067/scene/resources/2d/tile_set.cpp#L1429C18-L1429C30
